Integration tests for import/include

samdoran opened the pull request:
##### SUMMARY

We need more tests for `include/import`. This is them.

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Bugfix Pull Request


##### COMPONENT NAME
`include_role`
`include_task`

`import_play`
`import_task`
`import_role`

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
2.5
```

samdoran committed:
First pass at include_role tests

samdoran committed:
Reorganize test structure

Do all import and include tests in a single target.

bcoca commented on test/integration/targets/include_import/play/test_import_play.yml:
```diff
@@ -0,0 +1,6 @@
+- hosts: localhost
+  connection: local
```
connection: local should be redundant

samdoran replied:
Fixed

bcoca commented on test/integration/targets/include_import/role/test_import_role.yml:
```diff
@@ -0,0 +1,70 @@
+        where_am_i_defined: in the task
+
+    # FIXME Currently failing
+    # Not sure if this is suppsode to work
```
it should work

samdoran replied:
Currently failing. I'll leave it in but commented out for now.

bcoca commented on test/integration/targets/include_import/role/test_import_role.yml:
```diff
@@ -0,0 +1,70 @@
+
+    - name: Use variable for role name
+      include_role:
+        name: "{{ role_name }}"
```
this checks play level var, we also need to check host level ones

samdoran committed:
Build out more tests and test with linear and free strategy for each type

samdoran committed:
import_role tests

samdoran committed:
Set target host for play test

samdoran committed:
Basic import_playbook tests

samdoran committed:
Basic import/include_tasks tests

samdoran committed:
Add recursion test

samdoran committed:
import_playbook tests

samdoran committed:
Add import_playbook group_vars test

Issue #33177

samdoran committed:
Additional group_var tests for import_playbook

samdoran committed:
Enable recursion test

samdoran committed:
More work on import_tasks

abadger commented:
@sdoran could we merge this now and add more tests in a separate PR later?

abadger commented:
Sorry sdoran......  @samdoran, question was for you.

samdoran commented:
@abadger Yes, that would probably be wise. It already contains a good number of tests though I know we need more.

Let me use @bcoca's suggestion of tagging things in order to skip the failing tests for now rather than commenting them out. I'll push a commit shortly.

samdoran committed:
Run all tests via run.sh, improve conditional tests, add never tag

Add never tag to certain failining tests and explicitly skip it for now until #34104 is merged. Some tasks need to remain commented out because they cause the entire play to fail and aren't skippable by tags because they are syntax and/or variable checks.

Improve test for whether or not a role was run based on a conditional. Since the [import|include]_role does not register a variable, use a variable that is set inside a role as a canary test for whether or not the role was run.

samdoran commented:
These aren't exhaustive and I'm sure I've missed some things, but I believe these are good enough to merge for now.

abadger commented:
rebuild_merge

mattclay commented:
CI failure due to consistently "unstable" test on all platforms:

https://app.shippable.com/github/ansible/ansible/runs/48928/48/tests

samdoran committed:
Use a fail task rather than trigger a failure via bogus command

This should resolve the "unstable" test results.

samdoran commented:
Switched to using `fail` module. Hopefully that will correct the "unstable" tests.

mattclay commented on test/integration/targets/include_import/roles/role1/tasks/fail.yml:
```diff
@@ -0,0 +1,2 @@
+- fail:
+    msg: This command should always fail
```
The task name must contain `EXPECTED FAILURE` for this to pass CI.

samdoran replied:
Thanks Matt! I'll get that fixed.

samdoran committed:
Import tag tests

Fix bug it tests where validate tasks weren't run on tag tests.
Add tests for task import/include with tags.

jctanner commented:
@samdoran can you rebase? Was there also another test that was still broken? Can you point it out?

samdoran commented:
Latest failure was transient and unrelated. Just re-ran and it passed. I’ll rebase in a bit.

abadger commented:
rebuild_merge

samdoran commented:
@jctanner There are a few tests failing that are either commented out or just skipped.

[Variables from a role with `private: no` do not get passed to the play](https://github.com/ansible/ansible/pull/33418/files#diff-8d237e199e8dd6fa6e4ac39886bab593R74)
[Passing an entire dictionary of variables to a role](https://github.com/ansible/ansible/pull/33418/files#diff-8d237e199e8dd6fa6e4ac39886bab593R74)

samdoran committed:
Remove test for playbook group_var inheritance

abadger commented:
Merged to devel and cherrypicked for the 2.4.3 rc3 release.
